### PR TITLE
add fast convolution traits.

### PR DESCRIPTION
**Pull request by natsala13**

### CODE REVIEW

File: utils/pre_run.py [hunk 1]
```diff
@@ -91,7 +92,7 @@ def __init__(self):
         self.parser = parser
 
     def after_parse(self, args):
-        assert (not (args.skeleton | args.conv3 | args.joints_pool)) | args.skeleton & (args.conv3 ^ args.joints_pool)
+        assert (not (args.skeleton | args.conv3 | args.joints_pool | args.fast)) | args.skeleton & (args.conv3 ^ args.joints_pool ^ args.fast)
```
Comment (sigal-raab) [inline utils/pre_run.py]: change the term 'fast' to 'conv3fast'

File: models/gan.py [hunk 2]
```diff
@@ -380,13 +393,14 @@ def forward(self, input, style):
 
 
 class ToXYZ(nn.Module):
-    def __init__(self, in_channel, style_dim, upsample=True, blur_kernel=[1, 3, 3, 1],
+    def __init__(self, in_channel, style_dim, upsample=True, blur_kernel=(1, 3, 3, 1),
```
Comment (sigal-raab) [inline models/gan.py:396]: Why change from list to tuple

File: utils/traits.py [hunk 3]
```diff
@@ -245,7 +248,7 @@ def reshape_1D_kernel(self, kernel):
 
     @staticmethod
     def n_joints(entity):
-        return [len(parents) for parents in entity.parents_list] #[1, 2, 6, 10, 15]
+        return [len(parents) for parents in entity.parents_list]  # TODO: just return entity.n_edges
```
Comment (sigal-raab) [inline utils/traits.py]: class Joint does not have a member n_edges. The comment you deleted is an example of a possible return value. Please undelete it. You may add the text "Example" in its prefix for clarity.

**Review by sigal-raab — CHANGES_REQUESTED** [review]
Please update the README.md file as well

File: models/gan.py (hunk 2 above)
Comment (natsala13) [reply models/gan.py:396]: No real reason, it is better convention to pass non mutable arguments as default arguments.

File: models/gan.py [hunk 4]
```diff
@@ -257,37 +262,51 @@ def __init__(
...
+        #        * if demodulation is NOT applied, scale pushes the std of the output to be defined by style,
+        #        and removes the effects of the increasing std that happens due to convolutions
+
+        # multiply by fan_in and by predicted std
+        # the scale multiplies the E dimension (2) and the style multiplies the channel one. (256)
+        weight = self.scale * weight * style
+
+        if self.demod_obj == 'weights':  # scale weights s.t. output features' std will be 1
+            demod = torch.rsqrt(weight.pow(2).sum(self.skeleton_traits.norm_axis(weight), keepdims=True) + 1e-8)
+            # scaling factor: 1 / sqrt(sum(w^2)) for kernel and channels
```
Comment (sigal-raab) [inline models/gan.py:291]: this comment should be ABOVE the code line

File: models/gan.py [hunk 5]
```diff
@@ -257,37 +262,51 @@ def __init__(
...
             weight = weight * demod
 
         weight = weight.view(
-            (weight.shape[0] * weight.shape[1],) + weight.shape[2:]  # ==>  batch * out_channel_expanded, in_channel, kernel_height, kernel_width
+            (weight.shape[0] * weight.shape[1],) + weight.shape[2:]
+            # ==>  batch * out_channel_expanded, in_channel, kernel_height, kernel_width
```
Comment (sigal-raab) [inline models/gan.py]: this comment should be ABOVE the code line

**Review by sigal-raab — APPROVED** [review]

**Pull request merged**